File: plugins/uns/root.go
package uns

import (
	"fmt"
	"os"
	"sync/atomic"
)

// DefaultRoot is the first topic segment when a deployment chooses none.
const DefaultRoot = "colca"

// Version is the second topic segment of every topic this package understands.
const Version = "v1"

// RootEnv names the environment variable that sets the root for a process.
const RootEnv = "COLCA_TOPIC_ROOT"

var root atomic.Pointer[string]

// Root is the first topic segment of every record this process handles. All
// nodes of one tree share it: nodes do not translate between roots, so a record
// under another root is refused like any topic outside the namespace.
func Root() string {
	if r := root.Load(); r != nil {
		return *r
	}
	return DefaultRoot
}

// Prefix is how every topic starts: the root and the version with the trailing
// slash, for example "colca/v1/".
func Prefix() string { return Root() + "/" + Version + "/" }

// SetRoot changes the root for the whole process. It is meant to be called at
// start-up, before anything builds or accepts a topic. An invalid root is
// refused and the previous one stays.
func SetRoot(r string) error {
	if err := ValidRoot(r); err != nil {
		return err
	}
	root.Store(&r)
	return nil
}

// SetRootFromEnv sets the root from COLCA_TOPIC_ROOT and keeps the default
// when the variable is unset or empty. Binaries that build topics without a
// node config call it first.
func SetRootFromEnv() error {
	r := os.Getenv(RootEnv)
	if r == "" {
		r = DefaultRoot
	}
	if err := SetRoot(r); err != nil {
		return fmt.Errorf("%s: %w", RootEnv, err)
	}
	return nil
}

// ValidRoot reports why r cannot be a topic root. A root is one topic segment
// of at most 64 letters, digits, '-', '_' or '.', starting with a letter or a
// digit — which rules out MQTT wildcards, the reserved "$" space and the "_"
// that marks a contract.
func ValidRoot(r string) error {
	if r == "" || len(r) > 64 {
		return fmt.Errorf("topic root must be 1 to 64 characters, got %q", r)
	}
	for i, c := range r {
		letterOrDigit := c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' || c >= '0' && c <= '9'
		if letterOrDigit || i > 0 && (c == '-' || c == '_' || c == '.') {
			continue
		}
		return fmt.Errorf("topic root %q: only letters, digits, '-', '_' and '.' are allowed, starting with a letter or digit", r)
	}
	return nil
}

File: contracts/src/colca_data_contracts/root.py
"""The topic root: the first segment of every Colca topic.

Every node, service and SDK process of one tree uses the same root. It is
``colca`` unless ``COLCA_TOPIC_ROOT`` names another. The value is read each time
it is asked for, so a process started with a different environment, or a test
that overrides it, gets the root it asked for.
"""
import os
import re

DEFAULT_ROOT = "colca"
ROOT_ENV = "COLCA_TOPIC_ROOT"
VERSION = "v1"

# One topic segment: no MQTT wildcards, not the reserved "$" space, and not the
# leading "_" that marks a contract.
_VALID = re.compile(r"[A-Za-z0-9][A-Za-z0-9._-]{0,63}")


def topic_root() -> str:
    """The configured root, or ``colca``. Raises ``ValueError`` for an invalid value."""
    value = os.environ.get(ROOT_ENV, "")
    if not value:
        return DEFAULT_ROOT
    if not _VALID.fullmatch(value):
        raise ValueError(
            f"{ROOT_ENV}={value!r} is not a topic root: one segment of at most 64 letters, "
            "digits, '-', '_' and '.', starting with a letter or digit"
        )
    return value


def topic_prefix() -> str:
    """How every topic starts: the root and the version, for example ``colca/v1/``."""
    return f"{topic_root()}/{VERSION}/"
